### ISSUE TITLE
Binary response WFCatalog 404

### ISSUE THREAD

**Issue by Jollyfant** [EIDA/mediatorws#60]
Hi, I was getting a spooky response to my WFCatalog request. I can reproduce this for the `federator-testing.ethz.ch` too. It happens only for GFZ requests. It is because the URL for GFZ is still resolved to: `http://geofon.gfz-potsdam.de/eidaws/wfcatalog/alpha/version` and not version `1`. That page returns a 404 with content type HTML and is interpreted as binary. I asked GFZ to fix this but the error should probably be propagated

    Forby:Rune Mathijs$ curl -v "http://federator-testing.ethz.ch/eidaws/wfcatalog/1/query?station=MTSE&start=2016-01-01&end=2016-01-02&channel=BHZ"
    *   Trying 129.132.144.214...
    * TCP_NODELAY set
    * Connected to federator-testing.ethz.ch (129.132.144.214) port 80 (#0)
    > GET /eidaws/wfcatalog/1/query?station=MTSE&start=2016-01-01&end=2016-01-02&channel=BHZ HTTP/1.1
    > Host: federator-testing.ethz.ch
    > User-Agent: curl/7.54.0
    > Accept: */*
    >
    < HTTP/1.1 200 OK
    < Date: Thu, 29 Nov 2018 12:10:36 GMT
    < Server: Apache/2.4.29 (Ubuntu)
    < Transfer-Encoding: chunked
    < Content-Type: application/json
    <
    [eQ?n?0??j?$Z????-???
    V?H?!??kR?VH??Ύwvvu????r8?S?4?U])=?Xa$Q?K"D-m ??i_?Wf????히?KdET?l? K[`Ƈc>d?x2Y?qm??4[?s?Y??? Y?v?!???JM??Sn;SԢi?c9??B????k?w?????????,??
    -m??W7,[?ca-
                O?Q-C*5җ?B!vBi?QZ?=p?dJ??U???
    ?E?<ˀ?w?;Y?????EP?h)?ۘRb?O/Eĸ???EWv????,FYoB??N??`
    * Connection #0 to host federator-testing.ethz.ch left intact
    ???4??]

Which is 320 bytes in between the JSON brackets (e.g. the default Apache Not Found page)

**Comment by damb** [comment]
Confirmed. However, it is difficult (IMO **impossible** without analyzing the response content) to handle such a behaviour if the response code returned from the endpoint is **200**. That is why I rather consider it as a *configuration issue* on the endpoint side than a `eida-federator` bug.

Since `eida-federator` has to rely on standards (especially when specified in the according specifications i.e. https://www.orfeus-eu.org/documents/WFCatalog_Specification-v0.22.pdf) currently I do not see any reason to fix this issue. Also, `eida-federator` behaviour fully agrees with such specs. Unfortunately this implies for this special case forwarding malformed data.

To sum it up: `eida-federator` cannot completely fulfill the specs if the endpoints already don't do so.

**Comment by damb** [comment]
Closed. Reopen if necessary.

**Comment by andres-h** [comment]
> Confirmed. However, it is difficult (IMO impossible without analyzing the response content) to handle such a behaviour if the response code returned from the endpoint is 200.

But we return 404... Apart from the status code, it looks like the "Content-Encoding: gzip" header is not handled correctly.

Maybe you need something like: https://github.com/GEOFON/fdsnws_scripts/blob/afe54ff8f059a7090b95e5217aa935b1496bbd8f/fdsnwsscripts/fdsnws_fetch.py#L500

**Comment by Jollyfant** [comment]
Agreed but how about this? It's not used? Seems like the request defaults to `1` regardless of what is there.

https://github.com/EIDA/mediatorws/blob/1ecf61ed324392878c9a9b0edae49f8eff9b4a8b/eidangservices/settings.py#L73

**Comment by damb** [comment]
To resolve the confusion:

```
$ curl --output - "http://federator-testing.ethz.ch/eidaws/routing/1/query?station=MTSE&start=2016-01-01&end=2016-01-02&channel=BHZ&service=wfcatalog"
http://geofon.gfz-potsdam.de/eidaws/wfcatalog/alpha/query
EE MTSE -- BHZ 2016-01-01T00:00:00 2016-01-02T00:00:00
```
Hence, `stationlite` resolves correctly and `eida-federator` issues a request to `http://geofon.gfz-potsdam.de/eidaws/wfcatalog/alpha/query`.

@Jollyfant:
`eida-federator` relies on `stationlite`. The `EIDA_NODES` configuration dictionary is deprecated and probably will be removed in future (currently we use it for `eida-stationlite-harvest` and `eida-mediator`). So everything said about `http://geofon.gfz-potsdam.de/eidaws/wfcatalog/1/query` is not relevant here. Also, the HTTP response code of `http://geofon.gfz-potsdam.de/eidaws/wfcatalog/1/query` is **404**. It would be handled properly by `eida-federator` (though this is irrelevant, as I already mentioned).

@andres-h:
You are right that currently `Content-Encoding: gzip` is ignored. Neither https://www.orfeus-eu.org/documents/WFCatalog_Specification-v0.22.pdf nor http://www.fdsn.org/webservices/FDSN-WS-Specifications-1.1.pdf don't say anything about encoded responses.

However, [RFC7231](https://tools.ietf.org/html/rfc7231#section-3.1.2.2) specifies it as a valid header field. So either we go for the approach

https://github.com/GEOFON/fdsnws_scripts/blob/afe54ff8f059a7090b95e5217aa935b1496bbd8f/fdsnwsscripts/fdsnws_fetch.py#L500

or we have to consider the `Content-Encoding` header field for all responses returning from endpoints in order to decode incoming data properly.

**Comment by kaestli** [comment]
For accept-encoding request header and content-encoding of the response, see RFC 2616, chapter 14.3 (https://tools.ietf.org/html/rfc2616#page-102) : 
if you send an accept-encoding application/json header, you can expect uncompressed response.
If you don't sen an accept-encoding header, you basically must expect anything. As anything is poorly defined, it is better if the federator sends accept-encoding headers. they must start with application/json (as it is the default of the service), but may be extended to include also application/gzip, if this is desirable (it probably is). However, then, the service provider is still free in the decision whether to support compression or not.

**Comment by andres-h** [comment]
> if you send an accept-encoding application/json header, you can expect uncompressed response.

You are possibly mixing "Accept" (applies to "Content-Type", eg., "application/json") and "Accept-Encoding" (applies to "Content-Encoding", eg., "gzip").

If "Accept-Encoding" is missing, the server (nginx, apache) will not compress the data. The reason for the fdsnws_patch above was that some (corporate) proxies add "Accept-Encoding: gzip", but don't transparently decompress data as they would be supposed to. We observed that problem at KNMI.

Since XML compresses well, supporting compression would be generally a good idea.

**Comment by Jollyfant** [comment]
I'm confused as to where the binary junk is coming from then if GFZ is returning `404` and the federator is supposedly handling that properly.

**Comment by damb** [comment]
> I'm confused as to where the binary junk is coming from then if GFZ is returning 404 and the federator is supposedly handling that properly.

There is no `http://geofon.gfz-potsdam.de/eidaws/wfcatalog/1/query` and HTTP status code **404**.

`http://geofon.gfz-potsdam.de/eidaws/wfcatalog/alpha/query` i.e. GFZ (as routed by `stationlite`) returns HTTP status code 200 but with `Content-Encoding: gzip`. So, what you receive for
```
http://federator-testing.ethz.ch/eidaws/wfcatalog/1/query?station=MTSE&start=2016-01-01&end=2016-01-02&channel=BHZ
```
is the `eidaws-wfcatalog` data for `station=MTSE&start=2016-01-01&end=2016-01-02&channel=BHZ` but `gzip` encoded with leading and trailing JSON square brackets.
